### PR TITLE
#1

### PR DESCRIPTION
**Pull request by max-bazarov**
#1 enhancement

### CODE REVIEW

**Review by Yakov-Varnaev — CHANGES_REQUESTED** [review]

File: inventory/services.py [hunk 1]
```diff
@@ -16,5 +18,25 @@ class BaseMaterialService(Service):
...
+
+
+class ArchiveModelService:
+
+    def __init__(self, instance: Model) -> None:
+        self.instance = instance
+
+    def update(self, **kwargs: dict[str, type[Model]]) -> Model:
+        self.archive()
+        for field in funcy.omit(self.instance._meta.fields, 'id'):
```
Comment (Yakov-Varnaev) [inline inventory/services.py:33]: Does this work? Function omit works with mappings, not with sequences.

```
__class__.objects.create(**funcy.omit(kwargs, ['id']))
```

File: inventory/services.py [hunk 2]
```diff
@@ -16,5 +18,25 @@ class BaseMaterialService(Service):
...
 class MaterialDestroyService(ModelDestroyService):
     model = Material
+
+
+class ArchiveModelService:
+
+    def __init__(self, instance: Model) -> None:
+        self.instance = instance
+
+    def update(self, **kwargs: dict[str, type[Model]]) -> Model:
```
Comment (Yakov-Varnaev) [inline inventory/services.py:31]: kwargs are kwargs, there is no need to annotate it. More than that we keys are not Model classes, but values of any type

**Review by Yakov-Varnaev — CHANGES_REQUESTED** [review]

File: inventory/tests.py [hunk 3]
```diff
@@ -27,4 +27,26 @@ def setUpClass(cls):
             'price': Decimal('1.11'),
             'unit': MaterialUnits.GRAMMS.value
         }
+
         cls.instance = Material.objects.create(**cls.data)
+
+
+class TestMaterialArchiveService(TestCase, BaseArchiveServiceTests):
```
Comment (Yakov-Varnaev) [inline inventory/tests.py:34]: You can add this Base class to existing test.

File: core/tests.py [hunk 4]
```diff
@@ -22,3 +22,31 @@ def test_destroy(self):
...
+            f'{self.archive_service.__class__.__name__} archive method does not return instance id'
+        )
+
+    def test_update(self):
+        new_instance = self.archive_service(self.instance).update(**self.update_data)
+
+        assert isinstance(new_instance, self.model), (
+            f'{self.archive_service.__class__.__name__} update method does not return new instance'
+        )
+        assert self.instance.archived == True, (
```
Comment (Yakov-Varnaev) [inline core/tests.py:47]: Shouldn't we get instance from db rather than get it from test instance?

File: core/tests.py [hunk 5]
```diff
@@ -22,3 +22,31 @@ def test_destroy(self):
...
+        assert count == self.model.objects.count(), (
+            f'{self.archive_service.__class__.__name__} does not archive instance'
+        )
+        assert archived_id == self.instance.pk, (
+            f'{self.archive_service.__class__.__name__} archive method does not return instance id'
+        )
+
+    def test_update(self):
+        new_instance = self.archive_service(self.instance).update(**self.update_data)
+
```
Comment (Yakov-Varnaev) [inline core/tests.py:43]: I'd like to see next asserts as well:

- [ ] amount of instances increased
- [ ] new instance has fields of old instance

File: inventory/services.py [hunk 6]
```diff
@@ -28,12 +30,12 @@ class ArchiveModelService:
     def __init__(self, instance: Model) -> None:
         self.instance = instance
 
-    def update(self, **kwargs: dict[str, type[Model]]) -> Model:
+    def update(self, **kwargs) -> Model:
         self.archive()
-        for field in funcy.omit(self.instance._meta.fields, 'id'):
+        for field in self.instance._meta.fields:
             if field.name not in kwargs:
```
Comment (Yakov-Varnaev) [inline inventory/services.py:36]: If kwargs are empty we should raise exception, because it does not make sense to create same instance.

**Review by Yakov-Varnaev — CHANGES_REQUESTED** [review]

File: inventory/tests.py (hunk 3 above)
Comment (max-bazarov) [reply inventory/tests.py:34]: Didn't get to which existing test

File: inventory/services.py (hunk 6 above)
Comment (max-bazarov) [reply inventory/services.py:36]: @Yakov-Varnaev Should I create custom exception?

Comment (Yakov-Varnaev) [reply inventory/services.py:36]: This exception will be fine here
https://docs.djangoproject.com/en/4.1/ref/exceptions/#validationerror

File: inventory/tests.py (hunk 3 above)
Comment (Yakov-Varnaev) [reply inventory/tests.py:34]: Rename `TestMaterialCreateService` to `TestMaterialService` and add BaseArchiveServiceTest there. Btw, rename `BaseArchiveServiceTest**s**` to `BaseArchiveServiceTest`

File: core/tests.py [hunk 7]
```diff
@@ -22,3 +22,38 @@ def test_destroy(self):
...
+        )
+        assert archived_id == self.instance.pk, (
+            f'{self.archive_service.__class__.__name__} archive method does not return instance id'
+        )
+
+    def test_update(self):
+        count = self.model.objects.count()
+        new_instance = self.archive_service(self.instance).update(**self.update_data)
+
+        assert self.instance._meta.fields == new_instance._meta.fields, (
```
Comment (Yakov-Varnaev) [inline core/tests.py:44]: This always will be true. You should, check if all fields except for updated ones stay same.

E.g. we updated price. We should check that name of Material new instance is the same as old one.

**Review by Yakov-Varnaev — CHANGES_REQUESTED** [review]

File: core/tests.py [hunk 8]
```diff
@@ -22,3 +25,46 @@ def test_destroy(self):
...
+            f'{self.archive_service.__class__.__name__} does not archive instance'
+        )
+        assert archived_id == self.instance.pk, (
+            f'{self.archive_service.__class__.__name__} archive method does not return instance id'
+        )
+
+    def test_update(self):
+        count = self.model.objects.count()
+        new_instance = self.archive_service(self.instance).update(**self.update_data)
+        old_fields = [field.name for field in self.instance._meta.fields]
```
Comment (Yakov-Varnaev) [inline core/tests.py:46]:
```
unchanged_fields = [f.name for f in self.instance._meta.fields if f.name not in self.update_data]
assert all(getattr(self.instance, f) == getatter(new_instance, f) for f in unchanged_fields), (...)
```

**Review by Yakov-Varnaev — CHANGES_REQUESTED** [review]

File: core/tests.py [hunk 9]
```diff
@@ -22,3 +22,41 @@ def test_destroy(self):
...
+        count = self.model.objects.count()
+        new_instance = self.archive_service(self.instance).update(**self.update_data)
+        unchanged_fields = [f.name for f in self.instance._meta.f if f.name not in self.update_data]
+
+        are_same = all(
+            getattr(self.instance, f) == getattr(new_instance, f) for f in unchanged_fields
+        )
+        assert are_same, (
+            f'{self.archive_service.__class__.__name__} new instance data'
+            f'is not equal to old instance data')
```
Comment (Yakov-Varnaev) [inline core/tests.py:50]: Not the proper way to close braces.

File: core/tests.py [hunk 10]
```diff
@@ -22,3 +22,41 @@ def test_destroy(self):
...
+            f'{self.archive_service.__class__.__name__} does not archive instance'
+        )
+        assert archived_id == self.instance.pk, (
+            f'{self.archive_service.__class__.__name__} archive method does not return instance id'
+        )
+
+    def test_update(self):
+        count = self.model.objects.count()
+        new_instance = self.archive_service(self.instance).update(**self.update_data)
+        unchanged_fields = [f.name for f in self.instance._meta.f if f.name not in self.update_data]
```
Comment (Yakov-Varnaev) [inline core/tests.py:43]: `self.instance._meta.fields`*

File: core/services.py [hunk 11]
```diff
@@ -16,3 +21,30 @@ class ModelDestroyService(Service):
...
+
+
+class ArchiveService:
+
+    def __init__(self, instance: Model) -> None:
+        self.instance = instance
+
+    def update(self, **kwargs) -> Model:
+        self.archive()
+        if not kwargs:
```
Comment (Yakov-Varnaev) [inline core/services.py:33]: kwargs should be cleaned first.

Comment (Yakov-Varnaev) [reply core/services.py:33]: Omit all fields which are not model fields

**Review by Yakov-Varnaev — CHANGES_REQUESTED** [review]

File: core/services.py (hunk 11 above)
Comment (max-bazarov) [reply core/services.py:33]: @Yakov-Varnaev how to validate kwargs

Comment (Yakov-Varnaev) [reply core/services.py:33]:
```
kwargs = funcy.omit(kwargs, [f.name for f in self.instance._meta.fields])
```

File: core/services.py [hunk 12]
```diff
@@ -16,3 +21,31 @@ class ModelDestroyService(Service):
...
+            raise ValidationError('No data to update')
+        for field in self.instance._meta.fields:
+            if field.name not in kwargs:
+                kwargs[field.name] = getattr(self.instance, field.name)
+        new_instance = self.instance.__class__.objects.create(**funcy.omit(kwargs, 'id'))
+        return new_instance
+
+    def archive(self) -> int:
+        try:
+            self.instance.archived = True
```
Comment (Yakov-Varnaev) [inline core/services.py:44]: This line will never raise exception.
```
if not hasattr(self.instance, 'archived'):
     raise FieldDoesNotExist(...)
```

**Review by Yakov-Varnaev — CHANGES_REQUESTED** [review]

File: core/services.py (hunk 12 above)
Comment (max-bazarov) [reply core/services.py:44]: @Yakov-Varnaev Why it will never raise exception?

Comment (Yakov-Varnaev) [reply core/services.py]: Because you will just add new attr to instance. Creation might raise an exception, but not assigning

**Review by Yakov-Varnaev — APPROVED** [review]

**Pull request merged**